Add sycl_khr_free_function_commands extension (#644)
KhronosGroup/SYCL-Docs: adoc/extensions/sycl_khr_free_function_commands.adoc

requirements and the creation of [code]#event# objects.

The creation of [code]#event# objects may incur overheads that increase the
latency of submitting commands to devices, even if the [code]#event# object is
immediately discarded and never used.
Similarly, using a [code]#handler# can incur overheads even when there are no
dependent events or accessors registered.
Requiring developers to opt-in to these features is therefore expected to
improve the performance of many SYCL programs, by ensuring that SYCL developers
only pay the associated costs when necessary.

== Dependencies

This extension has no dependencies on other extensions.

== Feature test macro

An implementation supporting this extension must predefine the macro
[code]#SYCL_KHR_FREE_FUNCTION_COMMANDS# to one of the values defined in the
table below.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|Initial version of this extension.
|===

== Usage example

The example below rewrites the application from
<<subsec:example.sycl.application>> to demonstrate the usage of this extension.

[source,role=synopsis]
----
#include <iostream>
#include <sycl/sycl.hpp>

> using namespace sycl; // (optional) avoids need for "sycl::" before SYCL names.

Reviewer: Should we edit the example by using some requirement?

  // This does not block the host, but enforces dependencies on the device.
  khr::command_barrier(myQueue);

  // Launch an asynchronous kernel to compute matrix addition c = a + b.
  // Require that the scheduler create an event to track completion.
  range<2> local = { 2, 2 };
  range<2> global = { N, M };
  auto reqs = khr::requirements(khr::tracking());
  auto ev = khr::launch_grouped(myQueue, global, local, reqs, [=](nd_item<2> it) {
    size_t i = it.get_global_id(0);
    size_t j = it.get_global_id(1);
    size_t index = i * M + j;
    c[index] = a[index] + b[index];
  });

  // Wait for all three kernels to complete before accessing the results.
  // This blocks the host until all previous kernels have completed.
  ev->wait();

  std::cout << std::endl << "Result:" << std::endl;
  for (size_t i = 0; i < N; i++) {
    for (size_t j = 0; j < M; j++) {
      size_t index = i * M + j;
      // Compare the result to the analytic value.
      if (c[index] != i * (2 + 2014) + j * (1 + 42)) {
        std::cout << "Wrong value " << c[index] << " on element " << i << " "
                  << j << std::endl;
        exit(-1);
      }
    }
  }

> 

Reviewer 2: Can we add an example with buffer/accessors somewhere?

free functions".

_Effects_: Calls [code]#h.require# for each [code]#accessor# passed to the
constructor of this [code]#requirements# object.

'''

.[apititle]#requirements::register-events#
[source,role=synopsis,id=api:register-events]
----
void register-events(handler& h) const;
----

This function is exposition only.
It is shown only to help specify the effect of the functions below under "New
free functions".

_Effects_: Calls [code]#h.depends_on# for each [code]#event# or
[code]#std::vector<event># passed to the constructor of this
[code]#requirements# object.

'''

.[apititle]#requirements::register-kernel-bundle#
[source,role=synopsis,id=api:register-kernel-bundle]
----
void register-kernel-bundle(handler& h) const;
----

This function is exposition only.
It is shown only to help specify the effect of the functions below under "New
free functions".

_Effects_: Calls [code]#h.use_kernel_bundle# if a [code]#kernel_bundle# in
executable state was passed to the constructor of this [code]#requirements#
object and has no effect otherwise.

'''

.[apititle]#requirements::has-tracking#
[source,role=synopsis,id=api:has-tracking]
----

> bool has-tracking() const;

Reviewer 3: ```suggestion
bool has_tracking() const;
```

Reviewer 4: I think the dash is intended.  John named all the exposition-only functions with dashes to help emphasize that they are not part of the specified API.

----

This function is exposition only.
It is shown only to help specify the effect of the functions below under "New
free functions".

_Returns_: [code]#true# if this [code]#requirements# object was constructed with
a [code]#tracking# object with tracking enabled, and [code]#false# otherwise.

'''

.[apititle]#tracking# constructor
[source,role=synopsis,id=api:tracking-ctor]
----
namespace sycl::khr {

tracking(bool enabled=true);

}
----

_Effects_: Construct a [code]#tracking# object, representing a requirement that
a command must be submitted such that its state can be tracked via an
[code]#event# when [code]#enabled# is [code]#true#.

{note}If an [code]#event# is _not_ required, [code]#tracking(false)# should be
expected to introduce a small amount of overhead compared to providing no
[code]#tracking# requirement.{endnote}

== New free functions

=== Kernel launch

// Launch a basic parallel_for with a function object.
// New form of queue::parallel_for(range, ...)
.[apititle]#launch# (kernel function)
[source,role=synopsis,id=api:launch]
----
namespace sycl::khr {

template <typename KernelType, typename... Requirements>

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 5: I'm probably too tired to answer this well, but...

The problem with passing it by universal reference is that it becomes unclear which signature of `k` is being called, should it have more than one function call operator.  For instance:

```c++
struct MyKernel {
    template<typename... Ts>
    void operator()(Ts&&...) { /* ... */ }

    template<typename... Ts>
    void operator()(Ts&&...) const { /* ... */ }
};
```

And how many kernels would really benefit from being movable?  Are there issues with move-only types?

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 2: > And how many kernels would really benefit from being movable?

In our implementation - everything using SYCL special classes (accessors/streams/etc.) would benefit, as those have `std::shared_ptr` inside managing SYCL RT internal details.

> Are there issues with move-only types?

I believe there are. Move-only types can be trivially-copyable and those are device-copyable by definition. We won't be able to copy them at all.

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 4: In particular, do we diagnose an error in this case:

* The member variable is _not_ used inside the kernel, and
* The type of the member variable is not a legal kernel argument.

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 4: @aelovikov-intel can you clarify what the use case is which justifies this change?  When we talked earlier, I thought the use case was this:

* The application has a named kernel object where some member variables are not used by the kernel; and
* Those member variables that are not used by the kernel can be moved in a way that is more efficient than copying; and
* The application constructs the named kernel object as a temporary; and
* The implementation needs to make a copy of the kernel object (in DPC++ this can be the case when host tasks are also submitted to the same queue as the kernel).

However, you say above that DPC++ would see a benefit for member variables that are `accessor` or other SYCL types with common reference semantics.  This is a bit confusing because it seems unlikely that the user would have an `accessor` member variable that is *not* used inside the kernel.

Are you saying that the motivating use case is *any* member variable that can be efficiently moved, even if that member variable *is* used as a kernel argument?

At first, I thought that `accessor` would not benefit from "move" optimization because the compiler has special magic when it passes an `accessor` as a kernel argument, and it certainly wouldn't use the move-assignment operator when passing such an argument.  However, maybe you are saying that DPC++ needs to copy / move the entire kernel object (in some cases) that is unrelated to passing kernel arguments.  Is that the case?

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 4: I think we are not considering any change to `sycl::handler::single_task` in this PR.  However, we could consider a change to `sycl::khr::launch_task`.  There is no "handler" in this case, so I think your example would look something like this:

```
#include <sycl/sycl.hpp>

int main() {
  sycl::queue q;
  sycl::buffer<int, 1> b{1};
  sycl::accessor acc{b, sycl::read_write};
  sycl::khr::requirements reqs{acc};
  sycl::khr::launch_task(q, reqs, [=]{ acc[0] = 42; });

  q.wait();
  assert(host_accessor{b}[0] == 42);
}
```

I actually can't think how to write this where `acc` is a temporary object because it needs to be used in both the `reqs` constructor and also in the body of the kernel.  Is there a use case where perfect forwarding is useful?

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 4: Let's rewrite this using a named kernel object, so the operators are more clear.  Also, there's no such thing as `event::set_not_ready` or `event::set_ready`, so let's use real code.

```
#include <sycl/sycl.hpp>

struct MyKernel {
  sycl::accessor acc;
  MyKernel(syc::accessor<int> a) : acc(a) {}
  void operator()() const { acc[0] = 42; }
};

int main() {
  sycl::queue q;
  sycl::buffer<int, 1> b{1};

  sycl::event e = q.submit([&](sycl::handler& cgh) {
    cgh.host_task([=](){ sleep(1); });
  });

  {
    sycl::accessor acc{b, sycl::read_write};
    sycl::khr::requirements reqs{acc, e};
    sycl::khr::launch_task(q, reqs, MyKernel{acc});
  }

  // Somewhere internally GPU task above gets enqueued here.
  // User kernel has to be alive at that time, meaning the kernel 
  // above had to be copied/moved internally.
  q.wait();
  assert(host_accessor{b}[0] == 42);
}
```

I think you are saying that the implementation of `launch_task` will (likely) need to copy / move the temporary `MyKernel` object in the code above, correct?  This is because the SYCL runtime will (likely) need to implement the host task in software via a thread, and it can't enqueue the kernel to the device until after the host task completes.  This implies that the SYCL runtime has either taken a copy or moved the kernel object because the host task may complete long after the call to `launch_task` has returned.

Do I have that right?

Completing the analysis, the `MyKernel` object contains an `accessor` member variable, which can be moved more efficiently than it can be copied.  Therefore, there would be a performance advantage if the SYCL runtime could move the temporary `MyKernel` object rather than copying it.

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 2: > Do I have that right?

Yes!

> the MyKernel object contains an accessor member variable, which can be moved more efficiently than it can be copied.

Yes too.

> Therefore, there would be a performance advantage if the SYCL runtime could move the temporary MyKernel object rather than copying it.

Yes. In addition to that, user kernel maybe be non-copyable and only movable. In that case, not being able to move would mean we can't even compile the code.

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 5: Move-only kernels would be a feature request which does not compile now.  It also severely limits implementation techniques (possibly to the point of unimplementable), as you can only have one copy of a kernel across all devices.

For `host_task` specifically, maybe, as the member function already takes it by forwarding reference, although I'm not really a fan of having different rules for host vs. device.

> std::optional<event> launch(const queue& q, range<1> r, const requirements<Requirements...>& reqs, const KernelType& k); (1)

Reviewer 4: Move-only kernel objects seem like a niche case, but I don't think we need to rely on that in order to justify the change that @aelovikov-intel proposes.  Does the `accessor` example above provide enough justification?

@nliber you said this above:

> The problem with passing it by universal reference is that it becomes unclear which signature of `k` is being called, should it have more than one function call operator. For instance:
> 
> ```c++
> struct MyKernel {
>     template<typename... Ts>
>     void operator()(Ts&&...) { /* ... */ }
> 
>     template<typename... Ts>
>     void operator()(Ts&&...) const { /* ... */ }
> };
> ```

I think the example above is clear because [section 4.12.1.](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:interfaces.kernels.as.function-objects) "Defining kernels as named function objects" says:

> The `operator()` member function must be const-qualified ...

However, this bizarre case would still be unclear:

```c++
struct MyKernel {
    void operator()() const { /* ... */ }
    void operator()() const volatile { /* ... */ }
};
```

I guess we could clarify this by changing section 4.12.1 to say:

> The `operator()` member function must be const-qualified and not volatile-qualified ...

Aside from this issue that @nliber raised, are there other disadvantages to @aelovikov-intel's proposal?

= SYCL_KHR_FREE_FUNCTION_COMMANDS

This extension provides an alternative mechanism for submitting commands to a
device via free-functions that require developers to opt-in to the creation of

> [code]#event# objects.

Author: I've added something in 32d11f570526eae97ec8a3d9b40e3c4666d95f5c.

I don't think this needs to be a `{note}`, because the text I've added is in the overview of the extension.  But I'll convert it to a note if @gmlueck disagrees.

> [code]#event# objects.

Reviewer 4: Agreed.  Text in the Overview section is already non-normative, so it's not necessary to format it as a `{note}`.

}
----
_Effects (1)_: Equivalent to [code]#h.prefetch(ptr, numBytes)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { prefetch(h, ptr,
numBytes); })#.

'''

.[apititle]#mem_advise#
[source,role=synopsis,id=api:mem_advise]
----
namespace sycl::khr {

void mem_advise(sycl::handler& h, void* ptr, size_t numBytes, int advice); (1)

void mem_advise(sycl::queue q, void* ptr, size_t numBytes, int advice);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.mem_advise(ptr, numBytes, advice)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { mem_advise(h,
ptr, numBytes, advice); })#.

'''

=== Command and event barriers

.[apititle]#command_barrier#
[source,role=synopsis,id=api:command_barrier]
----
namespace sycl::khr {

void command_barrier(sycl::handler& h); (1)

void command_barrier(sycl::queue q);    (2)

}
----
_Effects_: Enqueues a command barrier.
Any commands submitted after this barrier cannot begin execution until:

* All commands previously submitted to this queue have completed; and
* All commands associated with this command's dependencies (e.g., via
  [code]#handler::depends_on#) have completed.

> 

Reviewer 6: Since `event_barrier` has a clarification for "no-op" case, shall we add one here too?

{note}Overload (1) is no-op when submitted to an in-order queue with no other dependencies (e.g., specified by [code]#handler::depends_on#). Overload (2) is no-op if [code]#queue# is in-order.{endnote}

Author: I've added a similar note in 165d07e13b7cfe379c44fe4da4544ed60ab05e69, but weakened from "is" to "may be" (for both notes) in be56bb77ae628ffafbcbd783c251b32aef51e70e.

Even though this is non-normative, upon reflection it seems a bad idea to guarantee that either of these things *is* a no-op, since that implies calling these functions does absolutely nothing and/or has no overhead. That might be true of some implementations, some of the time, but there will probably still be corner cases where _something_ has to happen (e.g., because the user requested an `event`, because the implementation missed an optimization opportunity, etc).

                           Reductions&&... reductions); (5)

template <typename KernelType, typename... Reductions>
void launch_grouped_reduce(queue q, range<3> r,
                           range<3> size, const KernelType& k,
                           Reductions&&... reductions); (6)

}
----
_Constraints_: The parameter pack consists of 1 or more objects created by the
[code]#reduction# function.

_Effects (1-3)_: Equivalent to [code]#+h.parallel_for(nd_range(r, size),
reductions..., k)+#.

_Effects (4-6)_: Equivalent to [code]#+q.submit([&](handler& h) {
launch_grouped_reduce(h, r, size, k, reductions...); })+#.

'''

.[apititle]#launch_task# (kernel function)
[source,role=synopsis,id=api:launch_task]
----
namespace sycl::khr {

template <typename KernelType>
void launch_task(handler& h, const KernelType& k); (1)

template <typename KernelType>
void launch_task(queue q, const KernelType& k);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.single_task(k)#.

_Effects (2)_: Equivalent to [code]#h.submit([&](handler& h) { launch_task(h,
k); })#.

'''

.[apititle]#launch_task# (kernel object)
[source,role=synopsis,id=api:launch_task-kernel]
----
namespace sycl::khr {

> template <typename Args...>

Reviewer 7: I think syntax is invalid here. It should be `template <typename... Args>`

Author: Thanks, I always get this backwards!  Fixed in d7ce65f434fbc141955a1d583c757547e7e92745.

void launch_task(handler& h, const kernel& k, Args&&... args); (1)

> template <typename Args...>

Reviewer 7: Same here `template <typename... Args>`

Author: Also fixed in d7ce65f434fbc141955a1d583c757547e7e92745.

template <typename SrcT, int SrcDims, access_mode SrcMode, typename DestT>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          DestT* dest);                 (3)

template <typename SrcT, int SrcDims, access_mode SrcMode, typename DestT>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          std::shared_ptr<DestT> dest); (4)

}
----

Copies from device to host.

_Constraints_:

* [code]#SrcT# and [code]#DestT# are <<device-copyable>>; and
* [code]#DestMode# is [code]#access_mode::read# or
  [code]#access_mode::read_write#.

_Preconditions_:

* [code]#dest# is a host pointer; and
* [code]#dest# points to an allocation of at least as many bytes as the range
  represented by [code]#src#.

_Effects (1-2)_: Equivalent to [code]#h.copy(src, dest)#.

_Effects (3-4)_: Equivalent to [code]#q.submit([&](handler& h) { copy(h, src,
dest); })#.

'''

.[apititle]#copy# (accessors, device to device)
[source,role=synopsis,id=api:copy-accessor-d2d]
----
namespace sycl::khr {

template <typename SrcT, int SrcDims, access_mode SrcMode,
          typename DestT, int DestDims, access_mode DestMode>

> void copy(sycl::queue q,

Reviewer 8: Likely a copy-paste typo.
```suggestion
void copy(sycl::handler& h,
```

Author: Good catch, thanks!

template <typename SrcT, int SrcDims, access_mode SrcMode,
          typename DestT, int DestDims, access_mode DestMode>
void copy(queue q,
          accessor<SrcT, SrcDims, SrcMode, target::device> src,
          accessor<DestT, DestDims, DestMode, target::device> dest); (1)

}
----

Copies between two device accessors.

_Constraints_:

* [code]#SrcT# and [code]#DestT# are <<device-copyable>>;
* [code]#SrcMode# is [code]#access_mode::read# or
  [code]#access_mode::read_write#; and
* [code]#DestMode# is [code]#access_mode::write# or
  [code]#access_mode::read_write#.

_Effects (1)_: Equivalent to [code]#h.copy(src, dest)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { h.require(src);
h.require(dest); copy(h, src, dest); })#.

_Throws_: A synchronous [code]#exception# with the [code]#errc::invalid# error
code if [code]#dest.get_count() < src.get_count()#.

'''

.[apititle]#memset#
[source,role=synopsis,id=api:memset]
----
namespace sycl::khr {

void memset(handler& h, void* ptr, int value, size_t numBytes); (1)

void memset(queue q, void* ptr, int value, size_t numBytes);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.memset(ptr, value, numBytes)#.

> _Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { memset(h, value,

Reviewer 7: I think we are lacking `ptr` as parameter. Corrected version: `q.submit([&](handler& h) { memset(h, ptr, value, numBytes); });`

Author: Good catch, thanks!  Fixed in 19440ec.

namespace sycl::khr {

void command_barrier(sycl::handler& h); (1)

void command_barrier(sycl::queue q);    (2)

}
----
_Effects_: Enqueues a command barrier.
Any commands submitted after this barrier cannot begin execution until:

* All commands previously submitted to this queue have completed; and
* All commands associated with this command's dependencies (e.g., via
  [code]#handler::depends_on#) have completed.

'''

.[apititle]#event_barrier#
[source,role=synopsis,id=api:event_barrier]
----
namespace sycl::khr {

void event_barrier(sycl::handler& h, const std::vector<sycl::event>& events); (1)

void event_barrier(sycl::queue q, const std::vector<sycl::event>& events);    (2)

}
----
_Effects_: Enqueues an event barrier.
Any commands submitted after this barrier cannot begin execution until:

* All commands associated with [code]#events# have completed; and
* All commands associated with this command's dependencies (e.g., via
  [code]#handler::depends_on#) have completed.

{note}For both overloads, if [code]#events# is empty and an event barrier has no
other dependencies (e.g., specified by [code]#handler::depends_on#), then this
operation is a no-op.{endnote}

'''

== Issues

> None.

Reviewer 4: Sorry, one more comment.  I propose that we omit the "Issues" section when there are no open issues, especially when we are ready to ratify a KHR.  For example, the "SYCL_KHR_DEFAULT_CONTEXT" specification does not have an "Issues" section.

Author: That makes sense.  Removed in fa8a8f6d60e5d3df80c29974b4083a4ea4aa9445.

device via free-functions that require developers to opt-in to the creation of
[code]#event# objects.

== Dependencies

This extension has no dependencies on other extensions.

== Feature test macro

An implementation supporting this extension must predefine the macro
[code]#SYCL_KHR_FREE_FUNCTION_COMMANDS# to one of the values defined in the
table below.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|Initial version of this extension.
|===

== Usage example

The example below rewrites the application from
<<subsec:example.sycl.application>> to demonstrate the usage of this extension.

[source,role=synopsis]
----
#include <iostream>
#include <sycl/sycl.hpp>
using namespace sycl; // (optional) avoids need for "sycl::" before SYCL names

// Size of the matrices
constexpr size_t N = 2000;
constexpr size_t M = 3000;

int main() {
  // Create a queue to work on
  queue myQueue;

  // Create some 2D arrays of float for our matrices
  float* a = malloc_shared<float>(N * M, myQueue);
  float* b = malloc_shared<float>(N * M, myQueue);
  float* c = malloc_shared<float>(N * M, myQueue);

  // Launch an asynchronous kernel to initialize a

>   // Use khr::submit to get a handler, even though not required here

Reviewer 9: What do you mean?

Author: The `handler` isn't strictly required here, because we don't use any functionality of the `handler` other than to enqueue the kernel.  But I agree the wording is confusing, so I've simplified it in ce08652764e915d0d5bb16628869a7e2f5efd657.  The wording is now more consistent with the usage below that doesn't use a `handler`.

  khr::submit(myQueue, [&](handler& cgh) {

    // Enqueue a kernel iterating on a N*M 2D iteration space
    khr::launch(cgh, range<2> { N, M }, [=](id<2> index) {
      size_t i = index[0];
      size_t j = index[1];
      a[i * M + j] = i * 2 + j;
    });

  });

>   // Launch an asynchronous kernel to initialize b

Reviewer 9: ```suggestion
  // Launch an asynchronous kernel to initialize b.
```
Use `.` at the end of each sentence so the comments are parseable.

Author: Done in 372bb3b2d8462248c96972b5ea69396c776f006a.

  // Use khr::launch to enqueue a kernel without a handler.
  khr::launch(myQueue, range<2> { N, M }, [=](id<2> index) {
    size_t i = index[0];
    size_t j = index[1];
    b[i * M + j] = i * 2014 + j * 42;
  });

  // Ensure that the two previous kernels complete before enqueueing more work.
  // This does not block the host, but enforces dependencies on the device.
  khr::command_barrier(myQueue);

  // Launch an asynchronous kernel to compute matrix addition c = a + b.
  // Use khr::launch_grouped to enqueue a kernel using groups without a handler.
  range<2> local = { 2, 2 };
  range<2> global = { N, M };

>   khr::launch_grouped(myQueue, global, local, [=](khr::invocation<2> ivc) {

Reviewer 4: Is the `invocation` type part of this KHR?  If so, I think it is underspecified.  The description of `launch_grouped` says that it is equivalent to `h.parallel_for(nd_range(r, size), k)`.  However, that's not true if kernel's call operator is expected to take an `invocation` instead of an `nd_item`.

Author: Ah, I think I was missing a constructor, which I've added in 16111b2d8a9e5ccd685d61a905dce9f158d094e8.

This way, `khr::launch_grouped` expects a lambda taking anything implicitly constructible from a `sycl::nd_item` (which is equivalent to the behavior of `parallel_for`).  The implementation automatically calls the `khr::invocation` constructor when it launches the kernel.

Reviewer 4: It seems like `khr::invocation` is orthogonal to the rest of this extension, then?  As you say, `handler::parallel_for` already allows the kernel parameter to be any type that is convertible from `nd_item`, so `khr::invocation` is also usable in kernels launched that way.

Why should `khr::invocation` be part of this extension rather than its own extension?  Or, maybe it should be part of "sycl_khr_group_interface"?  Is the main purpose of `khr::invocation` to align the naming with that extension?

Author: Good question...

I don't think there's any dependency here that would prevent us from splitting things out into a separate extension if that's what we want to do.  But if functionality _can_ be meaningfully grouped together, it might be better to have fewer extensions -- that means fewer macros for developers to check, and less chance of a perception that we introduced "too many extensions".

If I recall correctly, we added `khr::invocation` here because we were discussing the free-function commands when it was proposed.  `khr::launch_grouped` removes the need for any `nd_range` object, and somebody suggested that we could use that as an opportunity to remove `nd_item` too.  So, the proposal for the SYCL-Next interface is a `launch_grouped` accepting an `invocation` -- that would be my main justification for putting them in the same extension.

You're right that the naming is aligned with "sycl_khr_group_interface", but I think that's just a consequence of trying to make sure all these different concepts would be compatible if we choose to promote all the KHRs to SYCL-Next.

Reviewer 4: > If I recall correctly, we added `khr::invocation` here because we were discussing the free-function commands when it was proposed. `khr::launch_grouped` removes the need for any `nd_range` object, and somebody suggested that we could use that as an opportunity to remove `nd_item` too.

I understand the motivation for removing `nd_range` from the API.  There was no way to do list-initialization of an `nd_range`, so you couldn't do something like `launch_grouped({SIZE},{WGSIZE}}, ...)`.  Splitting it into two `range` parameters solves this, because you *can* do `launch_grouped({SIZE}, {WGSIZE}, ...)`.

I don't understand the motivation for renaming `nd_item` to `invocation`.  On the surface, this just seems like superfluous naming churn.  If the main motivation is just to clarify the naming, it seems like this should be part of some larger extension.  For example, if we don't like the name `nd_item`, do we still like the name `item`?  It seems weird to rename the one and not the other.

Author: As discussed offline, I'm convinced that we should continue discussion about potential replacements for `nd_item` elsewhere.  Prototyping and testing the rest of this KHR does not rely on a replacement for `nd_item`, so I've removed the `khr::invocation` class in 269706cd5c2255b77204cabdb75d6595a068baa2.

}
----
_Effects (1-3)_: Equivalent to [code]#h.parallel_for(r, k)#.

_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) { launch(h, r,
k); })#.

'''

.[apititle]#launch# (kernel object)
[source,role=synopsis,id=api:launch-kernel]
----
namespace sycl::khr {

template <typename... Args>
void launch(sycl::handler& h, sycl::range<1> r,
            const sycl::kernel& k, Args&&... args); (1)

template <typename... Args>
void launch(sycl::handler& h, sycl::range<2> r,
            const sycl::kernel& k, Args&&... args); (2)

template <typename... Args>
void launch(sycl::handler& h, sycl::range<3> r,
            const sycl::kernel& k, Args&&... args); (3)

template <typename... Args>
void launch(sycl::queue q, sycl::range<1> r,
            const sycl::kernel& k, Args&&... args); (4)

template <typename... Args>
void launch(sycl::queue q, sycl::range<2> r,
            const sycl::kernel& k, Args&&... args); (5)

template <typename... Args>
void launch(sycl::queue q, sycl::range<3> r,
            const sycl::kernel& k, Args&&... args); (6)

}
----
_Effects_: Enqueues a kernel object that is invoked for every work-item in the
specified [code]#sycl::range#.

> The arguments in [code]#args# are passed to the kernel in the same order.

Reviewer 4: The core spec lists a bunch of other semantics when launching a `kernel` object.  Rather than duplicating all the text here, can we say something like:

> _Effects (1) - (3):_ Equivalent to:
> ```
> h.set_args(args...);
> h.parallel_for(r, k);
> ```
> _Effects (4) - (6):_ Equivalent to `q.submit([&](handler& h) { launch(h, r, k, args...); })`

Same for the other APIs below that take a `kernel` object.

Author: I've tried to fix this in 437754958e8597148fe1479d7a73691197aa947e.  I'd appreciate it if you could double-check that I got it right.

template <typename KernelType, typename... Reductions>
void launch_reduce(sycl::handler& h, sycl::range<1> r,
                   const KernelType& k, Reductions&&... reductions); (1)

template <typename KernelType, typename... Reductions>
void launch_reduce(sycl::handler& h, sycl::range<2> r,
                   const KernelType& k, Reductions&&... reductions); (2)

template <typename KernelType, typename... Reductions>
void launch_reduce(sycl::handler& h, sycl::range<3> r,
                   const KernelType& k, Reductions&&... reductions); (3)

template <typename KernelType, typename... Reductions>
void launch_reduce(sycl::queue q, sycl::range<1> r,
                   const KernelType& k, Reductions&&... reductions); (4)

template <typename KernelType, typename... Reductions>
void launch_reduce(sycl::queue q, sycl::range<2> r,
                   const KernelType& k, Reductions&&... reductions); (5)

template <typename KernelType, typename... Reductions>
void launch_reduce(sycl::queue q, sycl::range<3> r,
                   const KernelType& k, Reductions&&... reductions); (6)

}
----
_Constraints_: The parameter pack consists of 0 or more objects created by the

> [code]#sycl::reduction# function.

Reviewer 4: Just checking that you really want to allow 0 reduction objects here.  Do we instead want to require at least 1?

Author: Good catch.  Fixed in 2527e90051562e2b28d4c446d35674c2cfbf83b8.

> _Effects (1-3)_: Equivalent to [code]#h.parallel_for(r, reductions..., k)#.

Reviewer 4: ```suggestion
_Effects (1-3)_: Equivalent to [code]#+h.parallel_for(r, reductions..., k)+#.
```

Minor editorial nit ... this will prevent the `...` from being rendered as an ellipsis (which we do elsewhere when `...` is part of a code snippet).  Same comment for the other places below where `...` appears in an inline `[code]` role.

Author: Fixed in 9747f7a4ad3deecda7a422be4ba89730c458da59.

----
namespace sycl::khr {

template <typename Args...>
void launch_task(sycl::queue q, const sycl::kernel& k, Args&&... args);    (1)

template <typename Args...>
void launch_task(sycl::handler& h, const sycl::kernel& k, Args&&... args); (2)

}
----
_Effects_: Enqueues a kernel object as a device task.
The arguments in [code]#args# are passed to the kernel in the same order.

'''

=== Memory operations

.[apititle]#memcpy#
[source,role=synopsis,id=api:memcpy]
----
namespace sycl::khr {

void memcpy(sycl::handler& h, void* dest, const void* src, size_t numBytes); (1)

void memcpy(sycl::queue q, void* dest, const void* src, size_t numBytes);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.memcpy(dest, src, numBytes)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { memcpy(h, dest,
src, numBytes); })#.

'''

.[apititle]#copy# (USM pointers)
[source,role=synopsis,id=api:copy-pointer]
----
namespace sycl::khr {

template <typename T>
void copy(sycl::handler& h, const T* src, T* dest, size_t count); (1)

template <typename T>
void copy(sycl::queue q, const T* src, T* dest, size_t count);    (2)

}
----

Copies between two USM pointers.

> _Constraints_: [code]#T# must be <<device-copyable>>.

Reviewer 4: ```suggestion
_Constraints_: [code]#T# is <<device-copyable>>.
```

I was looking recently at the C++ specification wording of _Constraints_, and they use the wording "is" instead of "must be".  For example, see the [containers library](https://eel.is/c++draft/containers).  I started adopting this wording style in #670 and the other "vec" PRs I drafted lately.

Author: Makes sense to me.  We should do a clean-up pass to get rid of all the "Available only if/when" constraints in SYCL 2020 and replace them with "is".

_Preconditions_: [code]#src# and [code]#dest# must be host pointers or USM

> pointers accessible on the device.

Reviewer 4: The core spec has this description (which is intended to be a precondition):

> The `dest` and `src` parameters must each either be a host pointer or a pointer within a USM allocation that is accessible on the handler’s device.

I think we should choose wording that is more similar to that because:

* I think it's clearer that `src` could be a host pointer while `dest` is a USM pointer or vice versa.
* It's clearer that the pointer need not be the value returned from a USM allocation.  They could instead be pointers inside of a larger USM allocation.

Also, this precondition from the core spec is missing from your description:

> If a pointer is to a USM allocation, that allocation must have been created from the same context as the handler’s queue.

Alternatively, we could somehow say that these APIs have the same preconditions and constraints as the member functions in the core spec.

[code]#src# and [code]#dest# must point to allocations of at least [code]#count#
elements of type [code]#T#.

_Effects (1)_: Equivalent to [code]#h.copy(src, dest, count)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { copy(h, src,
dest, count); })#

'''

.[apititle]#copy# (accessors, host to device)
[source,role=synopsis,id=api:copy-accessor-h2d]
----
namespace sycl::khr {

template <typename SrcT, typename DestT, int DestDims, access_mode DestMode>
void copy(sycl::handler& h,
          const SrcT* src,

>           sycl::accessor<DestT, DestDims, DestMode, target::device> dest); (3)

Reviewer 4: ```suggestion
          sycl::accessor<DestT, DestDims, DestMode, target::device> dest); (1)
```

Our style is to restart the numbering from (1) in each synopsis code block.

Author: Even when the code block is effectively a continuation of the one before?  These are all overloads of `copy`, and I only split them up in an attempt to keep things readable.

I can imagine there might be cases in future where we might want to say something like "Overload (3) of `copy`", and that seems simpler than saying "Overload (1) of `copy` in the second synopsis code block".

I don't feel strongly about this, I just wanted to make sure you saw that these are all `copy` overloads and talk that through before making the change.

Reviewer 4: If you feel that the APIs are so closely related that the numbering should be a continuation, then I would encourage you to list them all in the same synopsis.  The advantage of this is that there will be a single `[apidef]` for all the `sycl::copy` functions, and any occurrence of `[api]#copy#` will hyperlink to that synopsis.  By splitting the `copy` functions into separate synopses, you lose this automatic hyperlinking feature.

The style I have been gravitating to is to prefer a single synopsis for all overloads of a function.  For example, in the reformatted `queue` section, there is a single synopsis for all 8 overloads of `queue::copy`.

Author: I feel that they're related in that they're all overloads of the same function, but I think that there's enough dissimilarity in the overloads that splitting them up improves readability.  There are 12 overloads in total here, but it's really 4 groups of related overloads.

I've restarted the numbering at 1, as you originally suggested: 88b540aff2c79f9c2fba8b2cf0bfa917b39434dc.  Everything is now consistent within this KHR.

_Effects (7-8)_: Equivalent to [code]#h.copy(src, dest)#.

_Effects (9-10)_: Equivalent to [code]#q.submit([&](handler& h) { copy(h, src,
dest); })#.

'''

.[apititle]#copy# (accessors, device to device)
[source,role=synopsis,id=api:copy-accessor-d2d]
----
namespace sycl::khr {

template <typename SrcT, int SrcDims, access_mode SrcMode,
          typename DestT, int DestDims, access_mode DestMode>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          sycl::accessor<DestT, DestDims, DestMode, target::device> dest); (11)

template <typename SrcT, int SrcDims, access_mode SrcMode,
          typename DestT, int DestDims, access_mode DestMode>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          sycl::accessor<DestT, DestDims, DestMode, target::device> dest); (12)

}
----

Copies between two device accessors.

_Constraints_: [code]#SrcT# and [code]#DestT# must be <<device-copyable>>.
[code]#SrcMode# must be [code]#access_mode::read# or
[code]#access_mode::read_write#.
[code]#DestMode# must be [code]#access_mode::write# or

> [code]#access_mode::read_write#.

Reviewer 4: ISO C++ is not consistent about this, but they seem to use a bulleted list when there are many constraints.  For example, see [here](https://eel.is/c++draft/containers#flat.map.modifiers-19).  Again, I've been adopting this style in #670 and the related "vec" PRs.  I'm using a bulleted list and the word "and" like:

> _Constraints:_
> * `SrcT` and `DestT` are device-copyable,
> * `SrcMode` is `access_mode::read` or `access_mode::read_write`, and
> * `DestMode` is `access_mode::write` or `access_mode::read_write`.

Author: Ooh.  I like this much more; adopted in 47c08f8623d9be26f7116fd9da71fa42ba71aa02.

namespace sycl::khr {

void prefetch(sycl::handler& h, void* ptr, size_t numBytes); (1)

void prefetch(sycl::queue q, void* ptr, size_t numBytes);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.prefetch(ptr, numBytes)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { prefetch(h, ptr,
numBytes); })#.

'''

.[apititle]#mem_advise#
[source,role=synopsis,id=api:mem_advise]
----
namespace sycl::khr {

void mem_advise(sycl::handler& h, void* ptr, size_t numBytes, int advice); (1)

void mem_advise(sycl::queue q, void* ptr, size_t numBytes, int advice);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.mem_advise(ptr, numBytes, advice)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { mem_advise(h,
ptr, numBytes, advice); })#.

'''

=== Command and event barriers

.[apititle]#command_barrier#
[source,role=synopsis,id=api:command_barrier]
----
namespace sycl::khr {

void command_barrier(sycl::handler& h); (1)

void command_barrier(sycl::queue q);    (2)

}
----
_Effects_: Enqueues a command barrier.
Any commands submitted after this barrier cannot begin execution until all
previously submitted commands (and any commands associated with dependendent

> events) have completed.

Reviewer 4: I assume "dependent events" means events specified via `handler::depends_on`?  It might be clearer to make this more explicit:

> Any commands submitted after this barrier cannot begin execution until:
> * All commands previously submitted to this queue have completed, and
> * All commands associated with this command's dependencies (e.g. via `handler::depends_on`) have completed.

Author: Fixed in c24fb1303aae2d9310e6657db4b5fb955a52474c.

'''

.[apititle]#event_barrier#
[source,role=synopsis,id=api:event_barrier]
----
namespace sycl::khr {

void event_barrier(sycl::handler& h, const std::vector<sycl::event>& events); (1)

void event_barrier(sycl::queue q, const std::vector<sycl::event>& events);    (2)

}
----
_Effects_: Enqueues an event barrier.
Any commands submitted after this barrier cannot begin execution until all
commands associated with [code]#events# (and any commands associated with other

> dependent events) have completed.

Reviewer 4: Same comment here about clarify "dependent events"

Author: Also fixed in c24fb1303aae2d9310e6657db4b5fb955a52474c.

{note}For both overloads, if [code]#events# is empty and an event barrier has no

> other dependencies (e.g., specified by [code]#handler::depends_on#), it is ot

Reviewer 4: ```suggestion
other dependencies (e.g., specified by [code]#handler::depends_on#), it is not
```

Author: Good catch!  Fixed in 9c627921aa959a199a29ce246491c5b995fdef58.

> required to wait for any commands unless the [code]#queue# is in-order.{endnote}

Reviewer 4: I found this a little confusing because a barrier doesn't really wait for anything.  It prevents subsequent commands from starting until previous ones have finished.  Maybe it would be less confusing to say:

> If `events` is empty, and an event barrier has no other dependencies (e.g. specified by `handler::depends_on`), then this operation is a no-op.

Author: I agree stating it's a no-op is clearer, so I've made that change in 664a9129b3130c4cc24ea7d833bed5957eeea615.

But I think we should discuss the mental model for these functions.  The decision we make here could impact both naming and implementation.  I see two different ways to describe the semantics of these functions:

1. The `barrier` command is enqueued to the device, and the device begins executing the `barrier` command as soon as its dependencies are satisfied.  The execution of the `barrier` command is an operation which waits for all previous commands/events and then completes.  This is consistent with how other SYCL and C++ "barrier" operations work (i.e., something always waits for something else at a barrier).  A user would probably expect a command with these semantics to be in `info::event_command_status::running` status between its dependencies being satisfied and the conditions for barrier completion being satisfied.

2. The `barrier` command is enqueued to the device, but the device doesn't execute anything. The submission of the `barrier` just tells the runtime that it cannot start executing later commands until certain conditions have been satisfied.  This is consistent with how other SYCL and C++ "fence" operations work (i.e., they define ordering constraints that implementations must obey).  A user would probably expect a command with these semantics to be in `info::event_command_status::running` status for 0 time, reflecting that the device doesn't actually do anything.

I don't know which of these is closest to the behavior of real implementations.  But if you think it's closer to 2), then we should consider renaming these functions `command_fence` and `event_fence`.

Reviewer 4: Regarding the implementation, I think we should allow either (1) or (2).  I see (2) as an optimization of (1).  In  general, I think it should be permissible for the implementation to optimize away the "running" phase of any command if it doesn't need to be done, so there is never any guarantee that an event's status will be `running` at some point.  In any case, this isn't observable because there's no way to guarantee that the host calls `event::get_info` at just the right time to observe the `running` status.

Regarding the name, I'd defer to your judgement.  It seems to me, though, that the spec currently uses the word "barrier" to mean some sort of synchronization between work-item execution while it uses the word "fence" to mean observability of memory operations.  It seems like this command is closer in meaning to "barrier" than to "fence".

Author: Good point about the observability.

I'm not sure what my preference is regarding naming, and I could go either way.  Part of me even wonders if it should be `command_fence` and `event_barrier`!  I'll give it some more thought.

Reviewer 4: My original concern is addressed, so I'm marking this thread as "resolved".  If you want to let others comment on the naming, feel free to un-resolve it, or start a new comment just on the naming.

'''

== [code]#invocation# class template

The [code]#invocation# class template identifies an invocation of a kernel
function.

Instances of the [code]#invocation# class template are not user-constructible
and are passed as an argument to each invocation of a kernel function.

[source,role=synopsis]
----
namespace sycl::khr {

template <int Dimensions>
class invocation
{
 public:
  static constexpr int dimensions = Dimensions;

  id<Dimensions> id() const noexcept;
  size_t linear_id() const noexcept;

  range<Dimensions> range() const noexcept;

  group<Dimensions> get_work_group() const noexcept;

>   sub_group<Dimensions> get_sub_group() const noexcept;

Author: I don't think so.  And -- annoyingly -- ISO C++ doesn't seem to be consistent, so we can't really rely on precedent.

The convention I've tried to follow when sketching out these KHRs is to use `get_` as a syntactic hint that an implementation is either 1) doing more than just returning a member of a class; and/or 2) that the thing being returned conceptually exists outside of the class.

So `invocation::id()` and `invocation::range()` don't have the `get_` prefix because instantiating an `id` and a `range` within each `invocation` is likely to be a common implementation choice, and because these properties conceptually "belong to" [/handwave] each invocation.

`invocation::get_group()` and `invocation::get_sub_group()` have the `get_` prefix because implementations might have to look-up which group an invocation belongs to and construct a `group`/`sub_group` class, and because the groups themselves conceptually exist outside of the invocation.

Although we don't have a clear precedent to follow, this at least seemed closely aligned to how several ISO C++ classes behave.  For example, in `std::vector`, the pointer returned by `data()` "belongs to" the class, whereas the allocator returned by `get_allocator()` conceptually exists outside of the vector.

>   sub_group<Dimensions> get_sub_group() const noexcept;

Reviewer 10: > The convention I've tried to follow when sketching out these KHRs is to use get_ as a syntactic hint that an implementation is either 1) doing more than just returning a member of a class; and/or 2) that the thing being returned conceptually exists outside of the class.

I suspected something along those lines; I think it's a good way of handling it!

template <typename... Requirements>
class requirements {
public:
  // Only available if: all Requirements satisfy is_requirement_v
  requirements(Requirements... values);

  void register-accessors(handler& h) const;     // exposition only
  void register-events(handler& h) const;        // exposition only
  void register-kernel-bundle(handler& h) const; // exposition only
  bool has-tracking() const;                     // exposition only
};

class tracking {
public:
  tracking(bool enabled=true);
};

}
----

Each instance of a type listed below defines a specific scheduling requirement.
For each type, the [code]#is_requirement# type trait is specialized such that
[code]#is_requirement_v<T># returns [code]#true#.

* [code]#event#: The command must not begin executing until the event is
  complete.

* [code]#std::vector<event>#: The command must not begin executing until all
  events in the vector are complete.

* [code]#accessor#: The command must not begin executing until the
  [code]#buffer# associated with the [code]#accessor# can be accessed in a
  manner compatible with the specified [code]#access_mode#.

> * [code]#tracking#: The command must be submitted such that its state can be

Reviewer 11: ```suggestion
* [code]#tracking#: The command must be submitted such that its status can be
```

  tracked via an [code]#event# when the [code]#tracking# object is constructed
  with an [code]#enabled# value of [code]#true#.

* [code]#kernel_bundle<bundle_state::executable>#: The command must be submitted
  using a <<device image>> from the kernel bundle.

'''

.[apititle]#Default constructor#
[source,role=synopsis,id=api:requirements-ctor]
----
template <typename... Requirements>
requirements(Requirements... values);
----

_Constraints_:
* [code]#is_requirement_v# returns [code]#true# for each type in
  [code]#Requirements#; and

> * [code]#Requirements# contains at most one

Reviewer 11: Should we also have a constraint that there is at most one `tracking`, as these could have conflicting values?

Reviewer 4: Do we know how to implement the constraint to forbid multiple occurrences of the same requirement object?  In order for the constraint to work, I think the type system needs to somehow count the number of `tracking` objects.  How would we do that?

Author: I'm pretty sure you can do this with some simple metaprogramming.

I can think of a few different ways that would work.  An implementation could filter the parameter pack to make multiple tuples (e.g., one of all the `accessor` objects, one of all the `event` objects), and then just `static_assert` if there was more than one `tracking` object.  Or an implementation could use a dedicated fold expression to increment some counter for each `tracking` object that it sees, and `static_assert` if that counter is greater than 1.

Reviewer 4: Would it be easier to forbid duplicates when the `requirements` object is constructed?  (This is different from your current wording, where it is legal to construct a `requirements` with duplicates, but then place a constraint on the functions that take it.)  We should consider the complexity of the metaprogramming that is required in each case because that will affect compile time.

Author: These constraints are on the `requirements` object constructor, so the intent was to do exactly as you say.

We could collect some data here, but I don't think the complexity of the metaprogramming is very high, and I would hope that the overhead is minimal. The number of requirements is likely to be very low in the common case (e.g., a handful of accessors, tracking turned on, or a few dependent events).

Implementations already do this sort of thing and I haven't heard any complaints -- for example, I know that DPC++ uses a lot of metaprogramming to separate out different kinds of `reduction` objects.

Reviewer 4: Oh, sorry.  I was confusing this constraint with the constraints below on the new free functions.

  [code]#kernel_bundle<bundle_state::executable>#.

_Effects_: Constructs a [code]#requirements# object representing the set of
requirements specified via the [code]#values# parameter pack.

> _Remarks_: Unless otherwise specified, if an instance of a requirement appears

Reviewer 11: When we refer to an instance appearing more than once, does this mean objects which compare equals? So for example SYCL classes with reference semantics.

Author: Yes, I think so.  As far as I can tell, it's legal to call `h.require(acc)` twice with the same accessor, or to call `h.depends_on(e)` with the same event multiple times.  I don't see a reason to disallow that behavior here.

Do you think we need to clarify here what it means for requirements to be equivalent?

more than once in the [code]#values# parameter pack, the [code]#requirements#
object behaves as if it had only been specified once.

'''

.[apititle]#requirements::register-accessors#
[source,role=synopsis,id=api:register-accessors]
----
void register-accessors(handler& h) const; // exposition only
----

> _Effects_: Calls [code]#h.require# for each [code]#accessor# passed to the

Reviewer 11: ```suggestion
_Effects_: Equivalent to calling [code]#h.require# for each [code]#accessor# passed to the
```

Author: I'm not sure about this.  These are exposition-only already, so I don't see a problem with saying that they *do* call the `handler` function.  These functions only appear as part of larger "Equivalent to" blocks that already include other calls to `handler` member functions.

constructor of this [code]#requirements# object.

'''

.[apititle]#requirements::register-events#
[source,role=synopsis,id=api:register-events]
----
void register-events(handler& h) const; // exposition only
----

> _Effects_: Calls [code]#h.depends_on# for each [code]#event# or

Reviewer 11: ```suggestion
_Effects_: Equivalent to calling [code]#h.depends_on# for each [code]#event# or
```

[code]#std::vector<event># passed to the constructor of this
[code]#requirements# object.

'''

.[apititle]#requirements::register-kernel-bundle#
[source,role=synopsis,id=api:register-kernel-bundle]
----
void register-kernel-bundle(handler& h) const; // exposition only
----

> _Effects_: Calls [code]#h.use_kernel_bundle# if a [code]#kernel_bundle# in

Reviewer 11: ```suggestion
_Effects_: Equivalent to calling [code]#h.use_kernel_bundle# if a [code]#kernel_bundle# in
```

object and has no effect otherwise.

'''

.[apititle]#requirements::has-tracking#
[source,role=synopsis,id=api:has-tracking]
----
bool has-tracking() const; // exposition only
----

_Returns_: [code]#true# if this [code]#requirements# object was constructed with
a [code]#tracking# object with tracking enabled, and [code]#false# otherwise.

'''

.[apititle]#tracking# constructor
[source,role=synopsis,id=api:tracking-ctor]
----
namespace sycl::khr {

tracking(bool enabled=true);

}
----

_Effects_: Construct a [code]#tracking# object, representing a requirement that
a command must be submitted such that its state can be tracked via an
[code]#event# when [code]#enabled# is [code]#true#.

{note}If an [code]#event# is _not_ required, [code]#tracking(false)# should be
expected to introduce a small amount of overhead compared to providing no
[code]#tracking# requirement.{endnote}

== New free functions

=== Kernel launch

// Launch a basic parallel_for with a function object.
// New form of queue::parallel_for(range, ...)
.[apititle]#launch# (kernel function)
[source,role=synopsis,id=api:launch]
----
namespace sycl::khr {

> template <typename KernelType, typename... Requirements>

Reviewer 11: Why do we have separate overloads for each dimensionality rather than having a template parameter?

Author: This is a simplification designed to make the specification easier to read.

In SYCL 2020, these functions are all defined as being templated on `int Dims`, but there's a special note in [Section 4.9.4.2.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_parallel_for_invoke) saying that `parallel_for` must be invocable with braced initializer lists.  A compiler can't infer the `Dims` parameter from a braced initializer list, so all implementations today have to define overloads for each dimension.

We decided that for this extension we'd just expose the separate overloads.

template <typename... Args>
std::optional<event> launch(const queue& q, range<1> r, const kernel& k, Args&&... args);                                         (4)

template <typename... Args>
std::optional<event> launch(const queue& q, range<2> r, const kernel& k, Args&&... args);                                         (5)

template <typename... Args>
std::optional<event> launch(const queue& q, range<3> r, const kernel& k, Args&&... args);                                         (6)

}
----

_Constraints_ (1-3): [code]#Requirements# does not contain a
[code]#kernel_bundle#.

_Effects_ (1-3): Equivalent to: +
[source,sycl]
----
event ev = q.submit([&](handler& h) {
  reqs.register-events(h);
  reqs.register-accessors(h);
  h.set_args(args...);
  h.parallel_for(r, k);
});
return (reqs.has-tracking()) ? ev : std::nullopt;
----

_Effects_ (4-6): Equivalent to: [code]#+return launch(q, r, {}, k, args...);+#.

'''

// Launch a basic parallel_for with a function object and reductions.
// New form of parallel_for(range, reduction, ...)
.[apititle]#launch_reduce# (kernel function)
[source,role=synopsis,id=api:launch_reduce]
----
namespace sycl::khr {

> template <typename KernelType, typename... Requirements, typename... Reductions>

Reviewer 11: This is a minor point, but I see that for these overloads we have `Requirements` and then `Reductions`, but in the overloads above for taking a kernel object, we have `Args` and then `Requirements`, should we be consistent in having the variadic function arguments as the last template arg pack? Or is there a reason it's ordered in this way.

Author: There's no reason things are ordered this way.  I noticed the inconsistency myself, but figured there was a good chance we'd be reordering the arguments anyway and so things would have to change.

I wanted to collect some feedback about which order things should appear in.  Should the template parameters always be in the same order across different overloads (which would put `Requirements` at the end)?  Or should the template parameters always match the order of the function arguments?  Or something else?

template <typename... Args, typename... Requirements>
std::optional<event> launch_task(const queue& q, const requirements<Requirements>& reqs, const kernel& k, Args&&... args); (1)

template <typename... Args>
std::optional<event> launch_task(const queue& q, const kernel& k, Args&&... args);                                         (2)

}
----
_Constraints_ (1): [code]#Requirements# does not contain a
[code]#kernel_bundle#.

_Effects_ (1): Equivalent to: +
[source,sycl]
----
event ev = q.submit([&](handler& h) {
  reqs.register-events(h);
  reqs.register-accessors(h);
  h.set_args(args...);
  h.single_task(k);
});
return (reqs.has-tracking()) ? ev : std::nullopt;
----

_Effects_ (2): Equivalent to [code]#+return launch_task(q, {}, k, args...);+#.

'''

=== Memory operations

.[apititle]#memcpy#
[source,role=synopsis,id=api:memcpy]
----
namespace sycl::khr {

template <typename... Requirements>
std::optional<event> memcpy(const queue& q, void* dest, const void* src, size_t numBytes, const requirements<Requirements>& reqs = {});

}
----

_Effects_: Equivalent to: +
[source,sycl]
----
event ev = q.submit([&](handler& h) {
  reqs.register-events(h);
  reqs.register-accessors(h);

>   reqs.register-kernel-bundle(h);

Reviewer 11: I wonder if we want to introduce additional constraints here against having a requirement for a kernel bundle when enqueueing a memory command. If we're mapping to the existing API, a kernel bundle association can be made by the user and the impl would just ignore it, but with this API the processing of requirements is more programatic so we could restrict uses which don't make sense.

Author: Why would the implementation ignore it?  I don't see anything about that in SYCL 2020.

I can't think of a good use-case for it, but it seems like SYCL 2020 allows for a user to create an explicit kernel bundle associated with a `context`, change the way that it's built/linked, and then use it for these commands.  In some implementations, something like a `copy` might be implemented as a kernel under the covers, rather than a call to something like a DMA function.

@gmlueck - Since you wrote a lot of the `kernel_bundle` text, I'd like to hear your opinion here.  Does a `kernel_bundle` include kernels associated with commands like `copy` and `memcpy`?  I can see that there wouldn't be a way to get something like a `kernel_id` associated with these commands -- does that mean that we should treat them as being missing from the bundle?

Author: I've made most of these changes in f31095a7144012e73f9930d303d3eb49def2b296, fcd2ca03709921c3a740c110e91fd8ef3f8f4cda and 1ee96981e6531cebd87c7890a78b6f1e3c58f013.  I didn't add the constraint for `target::host_task`, because we don't have a host task yet.

One thing I noticed while working on 1ee96981e6531cebd87c7890a78b6f1e3c58f013, though, is that the `target::` parameter of an `accessor` seems pretty pointless now.  Device commands only accept `target::device` accessors, and host commands only accept `target::host_task` accessors.  It seems like we could probably deprecate this in the future, since the runtime can figure out how the `accessor` is going to be used based on context.

public:
  tracking(bool enabled=true);
};

}
----

Each instance of a type listed below defines a specific scheduling requirement.
For each type, the [code]#is_requirement# type trait is specialized such that
[code]#is_requirement_v<T># returns [code]#true#.

* [code]#event#: The command must not begin executing until the event is
  complete.

* [code]#std::vector<event>#: The command must not begin executing until all
  events in the vector are complete.

* [code]#accessor#: The command must not begin executing until the
  [code]#buffer# associated with the [code]#accessor# can be accessed in a
  manner compatible with the specified [code]#access_mode#.

* [code]#tracking#: The command must be submitted such that its status can be
  tracked via an [code]#event# when the [code]#tracking# object is constructed
  with an [code]#enabled# value of [code]#true#.

* [code]#kernel_bundle<bundle_state::executable>#: The command must be submitted
  using a <<device image>> from the kernel bundle.

'''

.[apititle]#Default constructor#
[source,role=synopsis,id=api:requirements-ctor]
----
template <typename... Requirements>
requirements(Requirements... values);
----

_Constraints_:

> * [code]#is_requirement_v# returns [code]#true# for each type in

Author: Fixed in 5edf35d2a5f37f674e8852295c5f2ac8f7b1d01a.

_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) { launch(h, r,
k); })#.

'''

.[apititle]#launch# (kernel object)
[source,role=synopsis,id=api:launch-kernel]
----
namespace sycl::khr {

template <typename... Args>
void launch(sycl::handler& h, sycl::range<1> r,
            const sycl::kernel& k, Args&&... args); (1)

template <typename... Args>
void launch(sycl::handler& h, sycl::range<2> r,
            const sycl::kernel& k, Args&&... args); (2)

template <typename... Args>
void launch(sycl::handler& h, sycl::range<3> r,
            const sycl::kernel& k, Args&&... args); (3)

template <typename... Args>
void launch(sycl::queue q, sycl::range<1> r,
            const sycl::kernel& k, Args&&... args); (4)

template <typename... Args>
void launch(sycl::queue q, sycl::range<2> r,
            const sycl::kernel& k, Args&&... args); (5)

template <typename... Args>
void launch(sycl::queue q, sycl::range<3> r,
            const sycl::kernel& k, Args&&... args); (6)

}
----
_Effects (1-3)_: Equivalent to:

[source,c++]
----
h.set_args(args...);
h.parallel_for(r, k);
----

_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) { launch(h, r, k,

> args...); })#.

Author: Fixed in 8fa1ce202f7ae9a4790ed27f386c8299f9656ab3.

'''

.[apititle]#launch_reduce# (kernel object)
[source,role=synopsis,id=api:launch_reduce-kernel]
----
namespace sycl::khr {

template <typename... Args>
void launch_reduce(sycl::handler& h, sycl::range<1> r,
                   const sycl::kernel& k, Args&&... args); (1)

template <typename... Args>
void launch_reduce(sycl::handler& h, sycl::range<2> r,
                   const sycl::kernel& k, Args&&... args); (2)

template <typename... Args>
void launch_reduce(sycl::handler& h, sycl::range<3> r,
                   const sycl::kernel& k, Args&&... args); (3)

template <typename... Args>
void launch_reduce(sycl::queue q, sycl::range<1> r,
                   const sycl::kernel& k, Args&&... args); (4)

template <typename... Args>
void launch_reduce(sycl::queue q, sycl::range<2> r,
                   const sycl::kernel& k, Args&&... args); (5)

template <typename... Args>
void launch_reduce(sycl::queue q, sycl::range<3> r,
                   const sycl::kernel& k, Args&&... args); (6)

}
----
_Effects (1-3)_: Equivalent to:
[source,c++]
----
h.set_args(args...);
h.parallel_for(r, k);
----

_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) {

> launch_reduce(h, r, k, args...); })#.

Author: `launch_reduce` performs a reduction and `launch` does not.

I don't know exactly how to specify this, though. My understanding is that the order in which arguments are passed to `sycl::kernel` launches is implementation-defined.  Do you think it would be sufficient to say something like:

> _Preconditions_:
> * The kernel performs 1 or more reductions; and
> * `Args` contains the required backend-specific arguments for reduction kernels.

> launch_reduce(h, r, k, args...); })#.

Author: Also fixed in https://github.com/KhronosGroup/SYCL-Docs/commit/8fa1ce202f7ae9a4790ed27f386c8299f9656ab3.

> launch_reduce(h, r, k, args...); })#.

Reviewer 4: We haven't defined any backend-specific arguments that correspond to the "reduction" object.  Therefore, I don't think it's possible to do backend interop with a "reduction" object.  The other use for `kernel` objects is the free-function kernel extension (an Intel extension), but free-function kernels do not support "reduction" objects either.  It's not clear to me that there is currently any use for this overload of `launch_reduce`.

> launch_reduce(h, r, k, args...); })#.

Author: I agree, but because the missing information is backend-specific, I think we need to define this overload in the specification.  Otherwise, there's no way for a backend to specify how to call a `kernel` object associated with a reduction. 

Any compiled kernel with a `reducer` argument knows what reduction is being performed, and `launch_reduce` knows how to launch that kernel and initialize all the data.  The only thing missing is what a user should pass to `set_args`, and I believe it's possible to define that (for a specific backend).

> launch_reduce(h, r, k, args...); })#.

Reviewer 4: > Otherwise, there's no way for a backend to specify how to call a `kernel` object associated with a reduction.

In the core spec, a `kernel` object can represent either:

* A device bulit-in kernel, or
* A kernel created from backend interop, for example creating a kernel from an OpenCL `cl_kernel`.

Therefore, the form of `launch` that takes a `kernel` is basically providing a way to launch a backend-specific kernel from SYCL.  A backend-specific kernel will never have an argument that is a "reduction" object because the reduction object is a SYCL-specific concept.

If that doesn't convince you, here is another argument.  Even if there was some way to create a `kernel` object that took a "reduction", the kernel would certainly have other non-reduction parameters.  Therefore, some of the arguments would be "reductions" and others would not.  I don't see why we would want a special function name for the case when some arguments are reductions.  We would instead expand the set of legal arguments for `handler::set_arg` in general, and then users could use the regular `launch` overload that takes a `kernel` object.

> launch_reduce(h, r, k, args...); })#.

Author: I don't think it's this simple.  Reductions are somewhat of a special case in SYCL, because there are quite a few implementations where a reduction is actually executed via multiple kernels.

In essence, you can view SYCL's reduction syntax as a shorthand for defining and launching a much more complicated sequence of kernels.  A SYCL kernel _never_ takes a `sycl::reduction` as an argument (note that the object is passed as an argument to `parallel_for`,  rather than being captured by the lambda), and by the time the user's lambda is converted into a kernel, it's been wrapped a few times.

The "real" reduction kernel often looks something like this:

```c++
template <typename UserProvidedLambda, typename Reduction>
struct RealReductionKernel {

  // This is the kernel function.
  void operator()(nd_item it) {
  
    // Initialize the reducer(s).
    reducer r = ...;
    
    // Launch the user-provided function.
    f(it, r);
    
    // Contribute results from the reducer(s) back to memory.
    *some_memory[it.get_group().get_group_id()] += r;

  }

  Reduction::T* some_memory;
  UserProvidedLambda f;

};
```

Calling `launch_reduce` for the implementation sketched above would actually launch a sequence of kernels:
1) Initialize `some_memory` to the identity of the reduction.
2) Call the reduction kernel derived from the user-provided lambda.
3) Call some additional kernels to combine all the work-group local results in `some_memory`.

Sorry for all the detail there, but I couldn't think of a shorter way to explain this, and need something to refer back to.

> A backend-specific kernel will never have an argument that is a "reduction" object because the reduction object is a SYCL-specific concept.

That's true, but as you can see above, the SYCL kernel doesn't have any "reduction" object arguments either.  The kernel enqueued by the backend just has a regular `T*` as an argument.  To define backend interop for reduction kernels, a backend would just need to define the expected interface for kernel 2).

> I don't see why we would want a special function name for the case when some arguments are reductions. We would instead expand the set of legal arguments for `handler::set_arg` in general, and then users could use the regular launch overload that takes a kernel object.

I think we could make this work, but I don't think the behavior would be as clear, and it might make things harder to implement.  What you're proposing would require `set_arg` to accept a `sycl::reduction` argument and convert that to a `T*` internally.  The regular `launch` overload taking a kernel object would then need to contain logic to check for reduction arguments, because it would need to conditionally generate calls to kernels 1) and 3).

By splitting the functions out, the regular `launch` can unconditionally assume it's just launching kernels in a straightforward manner, while `launch_reduce` can unconditionally assume it has to call kernels 1) and 3).  My understanding is that this is why @illuhad initially suggested this split, but Aksel please correct me if I'm wrong.

I 100% agree that nobody has defined behavior like this yet.  But since the purpose of this KHR is to gather feedback from users and implementers, I think it makes sense to give backends an opportunity to try and implement reduction interop instead of ruling it out now.

> launch_reduce(h, r, k, args...); })#.

Author: @gmlueck and I met and discussed this offline.  He convinced me that there's no reason for this function to exist until we have a situation where a backend actually decides to expose backend interop for reductions.  If that happens, we can always add these definitions back via a revision of the KHR.

I've removed the `launch_reduce` and `launch_grouped_reduce` overloads accepting a kernel object in d26831a0415669dc985a815e3fc89b6f1d34cabb.

                           sycl::range<2> size, const sycl::kernel& k,
                           Args&&... args); (2)

template <typename... Args>
void launch_grouped_reduce(sycl::handler& h, sycl::range<3> r,
                           sycl::range<3> size, const sycl::kernel& k,
                           Args&&... args); (3)

template <typename... Args>
void launch_grouped_reduce(sycl::queue q, sycl::range<1> r,
                           sycl::range<1> size, const sycl::kernel& k,
                           Args&&... args); (4)

template <typename... Args>
void launch_grouped_reduce(sycl::queue q, sycl::range<2> r,
                           sycl::range<2> size, const sycl::kernel& k,
                           Args&&... args); (5)

template <typename... Args>
void launch_grouped_reduce(sycl::queue q, sycl::range<3> r,
                           sycl::range<3> size, const sycl::kernel& k,
                           Args&&... args); (6)

}
----
_Effects (1-3)_: Equivalent to:
[source,c++]
----
h.set_args(args...);
h.parallel_for(nd_range(r, size), k);
----

_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) {

> launch_grouped_reduce(h, r, size, k, args...); })#.

Author: I think the solution here will be the same as for `launch` and `launch_reduce`, so let's discuss this in https://github.com/KhronosGroup/SYCL-Docs/pull/644#discussion_r1873393778.

> launch_grouped_reduce(h, r, size, k, args...); })#.

Author: Also fixed in https://github.com/KhronosGroup/SYCL-Docs/commit/8fa1ce202f7ae9a4790ed27f386c8299f9656ab3.

'''

.[apititle]#launch_task# (kernel function)
[source,role=synopsis,id=api:launch_task]
----
namespace sycl::khr {

template <typename KernelType>
void launch_task(sycl::handler& h, const KernelType& k); (1)

template <typename KernelType>
void launch_task(sycl::queue q, const KernelType& k);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.single_task(k)#.

_Effects (2)_: Equivalent to [code]#h.submit([&](handler& h) { launch_task(h,
k); })#.

'''

.[apititle]#launch_task# (kernel object)
[source,role=synopsis,id=api:launch_task-kernel]
----
namespace sycl::khr {

template <typename Args...>
void launch_task(sycl::handler& h, const sycl::kernel& k, Args&&... args); (1)

template <typename Args...>
void launch_task(sycl::queue q, const sycl::kernel& k, Args&&... args);    (2)

}
----
_Effects (1)_: Equivalent to:
[source,c++]
----
h.set_args(args...);
h.parallel_for(k);
----

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { launch_task(h, k,

> args...); })#.

Author: Also fixed in https://github.com/KhronosGroup/SYCL-Docs/commit/8fa1ce202f7ae9a4790ed27f386c8299f9656ab3.

----
namespace sycl::khr {

void memcpy(sycl::handler& h, void* dest, const void* src, size_t numBytes); (1)

void memcpy(sycl::queue q, void* dest, const void* src, size_t numBytes);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.memcpy(dest, src, numBytes)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { memcpy(h, dest,
src, numBytes); })#.

'''

.[apititle]#copy# (USM pointers)
[source,role=synopsis,id=api:copy-pointer]
----
namespace sycl::khr {

template <typename T>
void copy(sycl::handler& h, const T* src, T* dest, size_t count); (1)

template <typename T>
void copy(sycl::queue q, const T* src, T* dest, size_t count);    (2)

}
----

Copies between two USM pointers.

_Constraints_: [code]#T# is <<device-copyable>>.

_Preconditions_:

* [code]#src# is a host pointer or a pointer within a USM allocation that is
  accessible on the device.
* [code]#dest# is a host pointer or a pointer within a USM allocation that is
  accessible on the device.
* [code]#src# and [code]#dest# both point to allocations of at least
  [code]#count# elements of type [code]#T#.
* If either [code]#src# or [code]#dest# is a pointer is to a USM allocation,

>   that allocation was created from the same context as the handler's queue.

Author: Makes sense. Fixed in 2ba23942e349613db64e21f47f0077a52ca617f6.

_Effects (1-2)_: Equivalent to [code]#h.copy(src, dest)#.

_Effects (3-4)_: Equivalent to [code]#q.submit([&](handler& h) { copy(h, src,
dest) })#

'''

.[apititle]#copy# (accessors, device to host)
[source,role=synopsis,id=api:copy-accessor-d2h]
----
namespace sycl::khr {

template <typename SrcT, int SrcDims, access_mode SrcMode, typename DestT>
void copy(sycl::handler& h,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          DestT* dest);                 (1)

template <typename SrcT, int SrcDims, access_mode SrcMode, typename DestT>
void copy(sycl::handler& h,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          std::shared_ptr<DestT> dest); (2)

template <typename SrcT, int SrcDims, access_mode SrcMode, typename DestT>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          DestT* dest);                 (3)

template <typename SrcT, int SrcDims, access_mode SrcMode, typename DestT>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          std::shared_ptr<DestT> dest); (4)

}
----

Copies from device to host.

_Constraints_:

> * [code]#SrcT# and [code]#DestT# is <<device-copyable>>.

Author: Fixed in e382dbcd7f01d88b35faa2c359f2a234483d9633.

* [code]#DestMode# is [code]#access_mode::read# or
  [code]#access_mode::read_write#.

_Preconditions_:

* [code]#dest# is a host pointer.
* [code]#dest# points to an allocation of at least as many bytes as the range
  represented by [code]#src#.

_Effects (1-2)_: Equivalent to [code]#h.copy(src, dest)#.

_Effects (3-4)_: Equivalent to [code]#q.submit([&](handler& h) { copy(h, src,
dest); })#.

'''

.[apititle]#copy# (accessors, device to device)
[source,role=synopsis,id=api:copy-accessor-d2d]
----
namespace sycl::khr {

template <typename SrcT, int SrcDims, access_mode SrcMode,
          typename DestT, int DestDims, access_mode DestMode>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          sycl::accessor<DestT, DestDims, DestMode, target::device> dest); (1)

template <typename SrcT, int SrcDims, access_mode SrcMode,
          typename DestT, int DestDims, access_mode DestMode>
void copy(sycl::queue q,
          sycl::accessor<SrcT, SrcDims, SrcMode, target::device> src,
          sycl::accessor<DestT, DestDims, DestMode, target::device> dest); (1)

}
----

Copies between two device accessors.

_Constraints_:

> * [code]#SrcT# and [code]#DestT# is <<device-copyable>>.

Author: Fixed in e382dbcd7f01d88b35faa2c359f2a234483d9633.

code if [code]#dest.get_count() < src.get_count()#.

'''

.[apititle]#memset#
[source,role=synopsis,id=api:memset]
----
namespace sycl::khr {

void memset(sycl::handler& h, void* ptr, int value, size_t numBytes); (1)

void memset(sycl::queue q, void* ptr, int value, size_t numBytes);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.memset(ptr, value, numBytes)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { memset(h, value,
numBytes); })#.

'''

.[apititle]#fill#
[source,role=synopsis,id=api:fill]
----
namespace sycl::khr {

template <typename T>
void fill(sycl::handler& h, T* ptr, const T& pattern, size_t count); (1)

template <typename T, int Dims, access_mode Mode>
void fill(sycl::handler& h,
          sycl::accessor<T, Dims, Mode, target::device> dest,
          const T& src);                                             (2)

template <typename T>
void fill(sycl::queue q, T* ptr, const T& pattern, size_t count);    (3)

template <typename T, int Dims, access_mode Mode>
void fill(sycl::queue q,
          sycl::accessor<T, Dims, Mode, target::device> dest,
          const T& src);                                             (4)

}

> ----

Author: I think we could argue that it's not necessary in this case, but I'm happy to add it if you think we need it.

[ISO C++ says](https://eel.is/c++draft/structure.specifications#4) that if an _Effects_ is defined in terms of equivalence, all the _Constraints_ etc from the original function apply. The specification of `h.fill` already says `T` must be _device copyable_.  The only problem is that the original spec is not described in terms of _Constraints_...

> ----

Author: You're right.  Sorry, this version of the extension is now quite different from what we started with in the DPC++ version, so I've lost track of what's been written where.  Some of these things weren't originally written in terms of equivalence.

I've added the missing constraints in db380b4e051510952e31cd6a4bf7231d3954ec76, to be consistent.

_Effects (1)_: Equivalent to [code]#h.fill(ptr, pattern, count)#.

_Effects (2)_: Equivalent to [code]#h.fill(dest, src)#.

_Effects (3)_: Equivalent to [code]#q.submit([&](handler& h) { fill(h, ptr,
pattern, count); })#.

_Effects (4)_: Equivalent to [code]#q.submit([&](handler& h) { fill(h, dest,
src); })#.

'''

.[apititle]#update_host#
[source,role=synopsis,id=api:update_host]
----
namespace sycl::khr {

template <typename T, int Dims, access_mode Mode>
void update_host(sycl::handler& h, accessor<T, Dims, Mode, target::device> acc); (1)

template <typename T, int Dims, access_mode Mode>
void update_host(sycl::queue q, accessor<T, Dims, Mode, target::device> acc);    (2)

}
----

> _Constraints_: [code]#T# is <<device-copyable>>.

Reviewer 4: We do not need to say this here because `T` is the element type of `accessor` and `accessor` already has a requirement that its element type is device copyable.  Therefore, it would not be possible to even construct an `accessor` with a non-device-copyable element type.

Author: Good point.  Removed in a9bdc102f37d54ae1cbadb103b67f4c992f7df45.

void prefetch(sycl::queue q, void* ptr, size_t numBytes);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.prefetch(ptr, numBytes)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { prefetch(h, ptr,
numBytes); })#.

'''

.[apititle]#mem_advise#
[source,role=synopsis,id=api:mem_advise]
----
namespace sycl::khr {

void mem_advise(sycl::handler& h, void* ptr, size_t numBytes, int advice); (1)

void mem_advise(sycl::queue q, void* ptr, size_t numBytes, int advice);    (2)

}
----
_Effects (1)_: Equivalent to [code]#h.mem_advise(ptr, numBytes, advice)#.

_Effects (2)_: Equivalent to [code]#q.submit([&](handler& h) { mem_advise(h,
ptr, numBytes, advice); })#.

'''

=== Command and event barriers

.[apititle]#command_barrier#
[source,role=synopsis,id=api:command_barrier]
----
namespace sycl::khr {

void command_barrier(sycl::handler& h); (1)

void command_barrier(sycl::queue q);    (2)

}
----
_Effects_: Enqueues a command barrier.
Any commands submitted after this barrier cannot begin execution until:

* All commands previously submitted to this queue have completed; and
* All commands associated with this command's dependencies (e.g., via

>   `handler::depends_on`) have completed.

Reviewer 4: ```suggestion
  [code]#handler::depends_on#) have completed.
```

Check for other occurrences of backtick also.

Author: Oops.  Thanks.  Fixed in 2575901220de997cdc88635a60253c4b6f95b143.
